olp-cpp-sdk-authentication/src/SignInResultImpl.cpp

```diff
@@ -91,5 +101,70 @@ const std::string& SignInResultImpl::GetScope() const { return scope_; }
 
 bool SignInResultImpl::IsValid() const { return is_valid_; }
 
+std::string SignInResultImpl::ParseJwtToken(const std::string& token) {
+  const std::string::size_type first_dot = token.find_first_of('.');
+  if (std::string::npos == first_dot) {
+    OLP_SDK_LOG_ERROR_F(kLogTag, "Cannot parse ClientId. Wrong token format.");
+    return std::string();
+  }
+
+  const std::string jws_header_encoded = token.substr(0, first_dot);
+
+  std::vector<std::uint8_t> decoded_bytes;
+  if (!utils::Base64Decode(jws_header_encoded, decoded_bytes)) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "Cannot parse ClientId. Non-decodable token format");
+    return std::string();
+  }
+
+  std::string jws_header_decoded;
+  std::copy(decoded_bytes.begin(), decoded_bytes.end(),
+            std::back_inserter(jws_header_decoded));
+
+  rapidjson::Document doc;
+  doc.Parse(jws_header_decoded.c_str());
+
+  if (doc.HasParseError() || !doc.IsObject()) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "Cannot parse ClientId. Defective token format");
+    return std::string();
+  }
+
+  auto client_id_field = doc.FindMember("aid");
+  if (client_id_field == doc.MemberEnd() ||
+      !client_id_field->value.IsString()) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "Cannot parse ClientId. Field does not exist or is not "
+                        "a string json value");
+    return std::string();
+  }
+
+  const std::string result = client_id_field->value.GetString();
+
+  if (result.empty()) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "Cannot parse ClientId. Incomplete token format");
+  }
+
+  return result;
+}
+
+std::string SignInResultImpl::ParseClientIdFromToken(const std::string& token) {
+  if (token.empty()) {
+    return std::string();
+  }
```

Comment:
Log an error or a debug message? This case shouldn't happen because we parse the access token at line 60.